lightning-invoice/src/payment.rs

```diff
@@ -527,6 +677,46 @@ where
 	pub fn remove_cached_payment(&self, payment_hash: &PaymentHash) {
 		self.payment_cache.lock().unwrap().remove(payment_hash);
 	}
+
+	/// Given a [`PaymentHash`], this function looks up inflight path attempts in the payment_cache.
+	/// Then, it uses the path information inside the cache to construct a HashMap mapping a channel's
+	/// short channel id and direction to the amount being sent through it.
+	///
+	/// This function should be called whenever we need information about currently used up liquidity
+	/// across payments.
+	fn create_inflight_map(&self) -> HashMap<(u64, bool), u64> {
+		let mut total_inflight_map: HashMap<(u64, bool), u64> = HashMap::new();
+		// Make an attempt at finding existing payment information from `payment_cache`. If it
+		// does not exist, it probably is a fresh payment and we can just return an empty
+		// HashMap.
+		for payment_info in self.payment_cache.lock().unwrap().values() {
+			for path in &payment_info.paths {
+				if path.is_empty() { break };
+				// total_inflight_map needs to be direction-sensitive when keeping track of the HTLC value
+				// that is held up. However, the `hops` array, which is a path returned by `find_route` in
+				// the router excludes the payer node. In the following lines, the payer's information is
+				// hardcoded with an inflight value of 0 so that we can correctly represent the first hop
+				// in our sliding window of two.
+				let our_node_id: PublicKey = self.payer.node_id();
+				let reversed_hops_with_payer: Vec<(u64, PublicKey)> = core::iter::once((0u64, our_node_id)).chain(
+					path.split_last().unwrap().1.iter().map(|hop| (hop.fee_msat, hop.pubkey))).rev().collect();
```

Comment:
> Additionally, could possibly use `iter().skip(1).map(...)` instead of `split_last().unwrap().1.iter().map(...)`

I don't think this is equivalent since the former skips the first item in the collection, whereas I'm trying to drop the last element out of my collection

Reply:
Ah, right. I think I misread the docs, which are worded backwards wrt the returned tuple.

Given you are using `rev()` with the `chain()`, you can do `iter().rev().skip(1).map(...).chain(core::iter::once(...))`

Reply:
ooo yeah, that's clever.